presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,472 @@
+            execute(connection, cleanupSql);
+        }
+        catch (SQLException e) {
+            log.warn(e, "Failed to cleanup temporary table: %s", temporaryTable);
+        }
+    }
+
+    @Override
+    public PreparedStatement buildSql(ConnectorSession session, Connection connection, JdbcSplit split,
+            JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        // All remote table become default.schema.table
+        RemoteTableName remoteTableName = new RemoteTableName(Optional.empty(),
+                Optional.ofNullable(table.getSchemaName()), table.getTableName());
+        return new QueryBuilder(this).buildSql(session, connection, remoteTableName, table.getGroupingSets(), columns,
+                table.getConstraint(), split.getAdditionalPredicate(), tryApplyLimit(table.getLimit()));
+    }
+
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection,
+            JdbcTypeHandle typeHandle)
+    {
+        String jdbcTypeName = typeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new PrestoException(JDBC_ERROR, "Type name is missing: " + typeHandle));
+
+        Optional<ColumnMapping> mapping = getForcedMappingToVarchar(typeHandle);
+        if (mapping.isPresent()) {
+            return mapping;
+        }
+        // TODO use ru.yandex.clickhouse.domain.ClickHouseDataType
+        if (jdbcTypeName.equals("String")) {
+            if (typeHandle.getColumnSize().get() == 0) {
+                return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+            }
+        }
+        if (typeHandle.getJdbcType() == Types.DATE) {
+            return Optional.of(dateColumnMapping());
+        }
+        if (jdbcTypeName.equals("DateTime")) {
+            // TODO use timestamp
+            return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+        }
+        if (typeHandle.getJdbcType() == Types.TIMESTAMP) {
+            if (getUnsupportedTypeHandling(session) == CONVERT_TO_VARCHAR) {
+                return mapToUnboundedVarchar(typeHandle);
+            }
+            return Optional.empty();
+        }
+
+        if (typeHandle.getJdbcType() == Types.DECIMAL) {
+            int precision = typeHandle.getColumnSize().get();
+            if (precision > Decimals.MAX_PRECISION) {
+                int decimalDigits = typeHandle.getDecimalDigits()
+                        .orElseThrow(() -> new IllegalStateException("decimal digits not present"));
+                int scale = min(decimalDigits, getDecimalDefaultScale(session));
+                return Optional.of(decimalColumnMapping(createDecimalType(Decimals.MAX_PRECISION, scale),
+                        getDecimalRoundingMode(session)));
+            }
+        }
+        // TODO add explicit mappings
+        return legacyToPrestoType(session, connection, typeHandle);
```

Comment:
The `legacyToPrestoType` method is deprecated & exists only for legacy reasons. We do not need to and should not use it in new code.

```suggestion
        return Optional.empty();
```